### PR TITLE
x509_vfy.c: Restore rejection of expired trusted (root) certificate - backport to 1.1.1

### LINKED ISSUE #13471: OpenSSL 1.1.1h is no longer reporting expired root CAs when using "verify"
I've run into an issue with OpenSSL 1.1.1h when trying to verify a certificate chain that ends at an expired root certificate. The expiration check worked fine in OpenSSL 1.1.1g, properly reporting an expired root CA. However, with OpenSSL 1.1.1h, it reports that the verification succeeded even though the root certificate was expired.

Here's what I used to reproduce this issue:

Example root certificate saved as root.pem:

```
-----BEGIN CERTIFICATE-----
MIICzzCCAbegAwIBAgIUOOQ+yEDsApy5V50CVjbyRg+Jb54wDQYJKoZIhvcNAQEL
BQAwDzENMAsGA1UEAwwEcm9vdDAeFw0yMDExMTkyMzA1MjJaFw0yMDExMjAyMzA1
MjJaMA8xDTALBgNVBAMMBHJvb3QwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQDwtPzP8MvxlC7uH1acWI1gRt5IYJXYejLJ5aGdSTp7WpeDu7ZpSusMbwcs
L7Vxydo5KCnEn+HhWxiBnuwBp43RgNVrTIFHFIHP+nsE+xnKDp7OG/T1zIuyLH5r
iCulyJ6ss3V0T3bHkFzWP7eN3gFsznOeYy09pZ4oHqbUv8CGuBE5J8PxdFiGj0ya
0LGbT7/Rj9HoOzyPCXtnuAAxWsusD6fmDJOhK0bZkimYJkgR17IqrMYbGu1Lao+L
9f8ZARJsgOEzzoWvze3/V2RVnCjnL+ecwdiuvCrrazcBs1oUyGjcy4KjfLxF0jrS
AVuWJZW/k8FVVcLSX0AOPHsCNveXAgMBAAGjIzAhMA8GA1UdEwEB/wQFMAMBAf8w
DgYDVR0PAQH/BAQDAgEGMA0GCSqGSIb3DQEBCwUAA4IBAQAgMFXQ3JbCDWvVrNyf
LHCVETn04lVjKKdXR1aTHqglw0ecIfmUgkca6weby78iCvAE1BexkafZsEaeehob
a0myk40NyTU/iPsgVUm/KvGNzMUeFW4kHzFzu50ilEovE5W/jabQZzQCXG9O8A6r
pztots2WX53BeUqX2iMQvT2/LFPMUYKOwPv3toLdHVAMwuMSsmI8DRgQ8klMd4Sa
bsxJL7oXDPdE7LahRJgEvNqye/CdumiGthxhFod/+BdKKK9zruC43TPlxlYgG+EZ
vidiApo4gXhvTecZHS5bWMnHQ/9QbcQnIDd7Nqm2hkWCSgY1Yvph1ex2oNNotXK0
odXK
-----END CERTIFICATE-----
```

Example server certificate saved as server.pem:

```
-----BEGIN CERTIFICATE-----
MIICzjCCAbagAwIBAgIUHiCc6FlLb4cAmSk96GPLGNwlfe4wDQYJKoZIhvcNAQEL
BQAwDzENMAsGA1UEAwwEcm9vdDAeFw0yMDExMTkyMzI5MzNaFw0yMTExMjEyMzI5
MzNaMBExDzANBgNVBAMMBnNlcnZlcjCCASIwDQYJKoZIhvcNAQEBBQADggEPADCC
AQoCggEBALruu/SthRMZKpk8oCLA29NBxV361OQETZHuoliuru4IgEEL33b4XzFw
U0Ay06tPvPenYy4sznMzbwAz9DsjPGsBo1BNXqyNhd23k8cl4A+s/pwQNtK/WZoM
VR6s3klaWFMvhZY/oStWK/KVLuJZDPPuJNXBgE6gU24BRmb1sckQNILMqVb8qmJY
70O6iYbZHKs3c+t2P93IKbT9ownS3017pUBKiavpAYT3i4SHJF2A7gjmBoWBs9zp
aiJjhgqB5xyzuoeMYv+DovCOSz6katr8NB9MnLMI5pGCqt/F2cwFHmF4cngeD74w
Igp2VdSGc1B18V0bFS3kRkko3vFtLssCAwEAAaMgMB4wDAYDVR0TAQH/BAIwADAO
BgNVHQ8BAf8EBAMCA6gwDQYJKoZIhvcNAQELBQADggEBAC1psVLCrti+LmipwZ1J
v/ZE4B4GzBwnifRRkKT+CIJ+XYmmQlljVdnpjcs1bIQ11K7yQjMF5ZVOQBFHYo5h
zvRZUsMbs0NsxRYBhE2m9xxTMe4ybDZhhSlaJnt6HHWUYxtVr6NAPvgUlYkoIBhh
UlIR7BuyQwOJVburbNMc0UdBJGI71RfP/fsk50Cf12YcBKqfayeeLDj9dS71SmQ9
9do6YiOfB8wGFX+p+iKuu7I0iPJNmGtRacwVKPlDzldtrxrDapNEDFx0cFsvwkSC
WPQvuLhveCx7lZNqWTpVaHJShOEsJR65xdn8Jk2cSzpfLwO/felsfTNv1OraDvY0
MrI=
-----END CERTIFICATE-----
```

Summary of decoded root CA, intentionally kept as bare-bones as possible:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            38:e4:3e:c8:40:ec:02:9c:b9:57:9d:02:56:36:f2:46:0f:89:6f:9e
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = root
        Validity
            Not Before: Nov 19 23:05:22 2020 GMT
            Not After : Nov 20 23:05:22 2020 GMT      <-- Expired
        Subject: CN = root
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    <snip>
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:TRUE
            X509v3 Key Usage: critical
                Certificate Sign, CRL Sign
    Signature Algorithm: sha256WithRSAEncryption
         <snip>
```

Summary of decoded server cert:

```
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            1e:20:9c:e8:59:4b:6f:87:00:99:29:3d:e8:63:cb:18:dc:25:7d:ee
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: CN = root
        Validity
            Not Before: Nov 19 23:29:33 2020 GMT
            Not After : Nov 21 23:29:33 2021 GMT
        Subject: CN = server
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    <snip>
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment, Key Agreement
    Signature Algorithm: sha256WithRSAEncryption
         <snip>
```

Command to verify the certificate:

```
openssl verify --CAfile root.pem server.pem
```

Expected output, which was seen on OpenSSL 1.1.1g:

```
CN = root
error 10 at 1 depth lookup: certificate has expired
error server.pem: verification failed
```

Unexpected output seen on OpenSSL 1.1.1h:

```
server.pem: OK
```

I haven't completely tracked down the source of the problem, but it seems related to changes in the check_issued() function in x509/x509_vfy.c. In 1.1.1g this returns 1, but in 1.1.1h it returns 0, causing the internal_verify() function to stop after checking the server cert and not check the root cert.

### PR DESCRIPTION
**Pull request by DDvO**
The certificate path validation procedure specified in RFC 5280 does not include checking the validity period of the trusted (root) certificate.
Still it is common good practice to perform this check.
Also OpenSSL did this until version 1.1.1h, yet commit e2590c3a162eb118c36b09c2168164283aa099b4 accidentally killed it.

See also https://tools.ietf.org/html/rfc5280#section-6.1
and https://security.stackexchange.com/questions/27987/behavior-when-validating-an-expired-certificate-chain.

The current commit restores the previous behavior.
It also removes the cause of that bug, namely counter-intuitive design of the internal function `check_issued()`, 
which was complicated by checks that actually belong to some other internal function, namely `find_issuer()`.

Moreover, this commit adds a regression check and proper documentation of the root cert validity period check feature, 
which had been missing so far.

Fixes #13471.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by DDvO** [comment]
Once this PR is merged I'll forward-port this to master, fixing #13427.

**Comment by t8m** [comment]
> Once this PR is merged I'll forward-port this to master, fixing #13427.

We have policy of merging to master first.

**Comment by t8m** [comment]
Although this is fixing regression in 1.1.1h so this is kind of urgent.

**Comment by DDvO** [comment]
> > Once this PR is merged I'll forward-port this to master, fixing #13427.
> 
> We have policy of merging to master first.

I see, so here is the same fix for master: #13590.

**Comment by t8m** [comment]
The test for the expired root cert needs to be added here as well.

**Comment by t8m** [comment]
Maybe also add Changes entry (which is not needed on the master) about this fixing a regression in 1.1.1h.

**Comment by DDvO** [comment]
> The test for the expired root cert needs to be added here as well.

The test for the expired root cert needs to be added here as well.
Done; I've also taken over from #13590 further related minor (doc) improvements.

**Comment by DDvO** [comment]
> Maybe also add Changes entry (which is not needed on the master) about this fixing a regression in 1.1.1h.

Done; for clarity I added a entry for both 1.1.1h (documenting the accidental change) and for 1.1.1i (documenting the restoration).

**Comment by DDvO** [comment]
We better put a hold on this until #13586, which restores Linux-based CI checks, is merged and this PR is rebased on it.

File: crypto/x509/x509_vfy.c [hunk 1]
```diff
@@ -331,30 +349,13 @@ static X509 *find_issuer(X509_STORE_CTX *ctx, STACK_OF(X509) *sk, X509 *x)
...
 }
 
-/*
- * Check that the given certificate 'x' is issued by the certificate 'issuer'
- * and the issuer is not yet in ctx->chain, where the exceptional case
- * that 'x' is self-issued and ctx->chain has just one element is allowed.
- */
-static int check_issued(X509_STORE_CTX *ctx, X509 *x, X509 *issuer)
+/* Check that the given certificate 'x' is issued by the certificate 'issuer' */
+static int check_issued(ossl_unused X509_STORE_CTX *ctx, X509 *x, X509 *issuer)
```
Comment (t8m) [inline crypto/x509/x509_vfy.c:353]: We don't use ossl_unused anywhere except headers in 1.1.1 (this is different in master) so maybe we should not introduce it here?

Comment (DDvO) [reply crypto/x509/x509_vfy.c]: I see, so removed `ossl_unused` here.

**Comment by DDvO** [comment]
> We better put a hold on this until #13586, which restores Linux-based CI checks, is merged and this PR is rebased on it.

Rebased on the meanwhile merged #13586. 
Now let's see how the Linux CI runs go; the Windows-based ones were fine already.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
This needs to be rebased to fix the conflict in CHANGES. I do not think this needs to restart the 24h grace period.

**Comment by DDvO** [comment]
> This needs to be rebased to fix the conflict in CHANGES. I do not think this needs to restart the 24h grace period.

Thx for pointing this out; done.

**Review by t8m — APPROVED** [review]
Still approved

**Comment by DDvO** [comment]
Merged - thanks @t8m

**Pull request closed without merging**